### ISSUE TITLE
[PyTorch]: bad digamma float32 accuracy

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#6190]
PyTorch master@92a0f78

Our digamma single-precision floating point accuracy is bad near the poles. This is not caught by our tests because we only test on a few values:

https://github.com/pytorch/pytorch/blob/92a0f7835e7e2f6342fe2fd32a25299ea86aed31/test/test_torch.py#L272-L274

```python
input = torch.tensor([-1.99999994], dtype=torch.float32)
fp32 = input.digamma()
fp64 = input.double().digamma()
print(((fp32.double() - fp64) / fp64).item())  # 0.24012388522631392
```

This is a relative error of **24%**

We also are returning real numbers at the poles (for both float32 and float64) when we should be returning inf.

**Comment by zou3519** [comment]
I think our current implementation approximates the derivative. I have a high-precision digamma function here: https://github.com/pytorch/pytorch/pull/3955 that I'll rebase soon.

**Comment by zou3519** [comment]
It looks like the problem is that float `tan` is too imprecise:
```
 torch.tan(torch.tensor(-1.99999994 * pi, dtype=torch.double)).item()
 1.884955598503599e-07

 torch.tan(torch.tensor(-1.99999994 * pi)).item()
 3.019916050561733e-07
```
